onnxruntime/core/providers/webgpu/tensor/concat.cc

```diff
@@ -0,0 +1,143 @@
+     << "      return i;" << std::endl
+     << "    }" << std::endl
+     << "  }" << std::endl
+     << "  return " << input_count << ";" << std::endl
+     << "}" << std::endl;
+  return ss.str();
+}
+
+const std::string AssignOutput(const std::vector<std::unique_ptr<ShaderVariableHelper>>& inputs, const ShaderVariableHelper& output, size_t input_index) {
+  std::ostringstream ss;
+  const auto& input = inputs[input_index];
+  ss << output.SetByOffset("global_idx", input->GetByIndices("indices")) << ";" << std::endl;
+  return ss.str();
+}
+
+const std::string AppendAssignOutputDataFunction(const std::vector<std::unique_ptr<ShaderVariableHelper>>& inputs, const ShaderVariableHelper& output) {
+  std::ostringstream ss;
+  size_t input_count = inputs.size();
+  ss.imbue(std::locale::classic());
+  ss << "fn assign_output_data(global_idx: u32, input_index: u32, indices: output_indices_t) {" << std::endl;
+  if (input_count == 0) {
+    ss << AssignOutput(inputs, output, 0);
+  } else {
+    for (size_t i = 0; i < input_count; ++i) {
+      if (i == 0) {
+        ss << "  if (input_index == 0u) {" << std::endl;
+      } else if (i == input_count - 1) {
+        ss << "  } else {" << std::endl;
+      } else {
+        ss << "  } else if (input_index == " << i << "u) {" << std::endl;
+      }
+      ss << "     " << AssignOutput(inputs, output, i);
+    }
+    ss << "  }" << std::endl;
+  }
+  ss << "}" << std::endl;
+  return ss.str();
+}
+Status ConcatProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  for (size_t i = 0; i < input_count_; ++i) {
+    shader.AddInput("input_" + std::to_string(i), ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
+  }
+  const auto& inputs = shader.GetInputVars();
+  const auto& output = shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
+  shader.AppendImplementation(AppendCalCulateInputIndexFunction(input_count_));
+  shader.AppendImplementation(AppendAssignOutputDataFunction(inputs, output));
+  shader.SetMainFunctionBody(shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size"),
+                             "  var indices = ", output.OffsetToIndices("global_idx"), ";\n",
+                             "  let indices_axis = ", output.IndicesGet("indices", axis_), ";\n",
+                             "  let input_index = calculate_input_index(indices_axis);\n",
+                             "  if (input_index != 0u) {\n",
+                             "     ", output.IndicesSet("indices", axis_, "indices_axis - uniforms.size_in_concat_axis[input_index - 1]"), ";\n",
+                             "  }\n",
+                             "  assign_output_data(global_idx, input_index, indices);\n");
+  return Status::OK();
+}
+
+Status Concat::ComputeInternal(ComputeContext& context) const {
+  auto input_count = context.InputCount();
+  InlinedTensorsVector input_tensors;
+  input_tensors.reserve(input_count);
```

Comment:
add a TODO to deal with `input_count + 1 /* output_count */ > max_allowed_buffer_per_shader`